waspc/src/Wasp/Project/Analyze.hs

```diff
@@ -69,8 +83,103 @@ analyzeWaspProject waspDir options = do
   where
     fileNotFoundMessage = "Couldn't find the *.wasp file in the " ++ toFilePath waspDir ++ " directory"
 
-analyzeWaspFile :: Psl.Schema.Schema -> Path' Abs File' -> IO (Either [CompileError] [AS.Decl])
-analyzeWaspFile prismaSchemaAst waspFilePath = do
+data WaspFile
+  = WaspLang !(Path' Abs (File WaspLangFile))
+  | WaspTs !(Path' Abs (File WaspTsFile))
+
+data WaspLangFile
+
+data WaspTsFile
+
+data CompiledWaspJsFile
+
+data SpecJsonFile
+
+analyzeWaspFile :: Path' Abs (Dir WaspProjectDir) -> Psl.Schema.Schema -> WaspFile -> IO (Either [CompileError] [AS.Decl])
+analyzeWaspFile waspDir prismaSchemaAst = \case
+  WaspLang waspFilePath -> analyzeWaspLangFile prismaSchemaAst waspFilePath
+  WaspTs waspFilePath -> analyzeWaspTsFile waspDir prismaSchemaAst waspFilePath
+
+analyzeWaspTsFile :: Path' Abs (Dir WaspProjectDir) -> Psl.Schema.Schema -> Path' Abs (File WaspTsFile) -> IO (Either [CompileError] [AS.Decl])
+analyzeWaspTsFile waspProjectDir _prismaSchemaAst waspFilePath = runExceptT $ do
+  -- TODO: I'm not yet sure where tsconfig.node.json location will come from
+  -- because we also need that knowledge to generate a TS SDK project.
+  compiledWaspJsFile <- ExceptT $ compileWaspTsFile waspProjectDir [relfile|tsconfig.node.json|] waspFilePath
+  specJsonFile <- ExceptT $ executeMainWaspJsFile waspProjectDir compiledWaspJsFile
+  contents <- ExceptT $ readDeclsJsonFile specJsonFile
+  liftIO $ putStrLn "Here are the contents of the spec file:"
+  liftIO $ print contents
+  return []
+
+compileWaspTsFile ::
+  Path' Abs (Dir WaspProjectDir) ->
+  Path' (Rel WaspProjectDir) File' ->
+  Path' Abs (File WaspTsFile) ->
+  IO (Either [CompileError] (Path' Abs (File CompiledWaspJsFile)))
+compileWaspTsFile waspProjectDir tsconfigNodeFileInWaspProjectDir waspFilePath = do
+  chan <- newChan
+  (_, tscExitCode) <-
+    concurrently
+      (readJobMessagesAndPrintThemPrefixed chan)
+      ( runNodeCommandAsJob
+          waspProjectDir
+          "npx"
+          [ "tsc",
+            "-p",
+            toFilePath (waspProjectDir </> tsconfigNodeFileInWaspProjectDir),
+            "--noEmit",
+            "false",
+            "--outDir",
+            toFilePath outDir
+          ]
```

Comment:
Aren't the file names `main.wasp.mts` and `main.wasp.mjs`?

The validation PR will help here since it enables to required certain values in `tsconfig.json` (which we will need to expand to three different `tsconfig.json` files).

Reply:
> Aren't the file names main.wasp.mts and main.wasp.mjs?

You can ignore that, it's a minor detail. I mostly use `ts` and `js` as canonical names for this file when I'm writing about other ideas.

Reply:
It seems to me like you put effor tinto explaining the situation here, but I have to admit I still didn't get most of it. I am certainly missing some context + have less knowledge of the ts toolchain, but I still feel like I should have been able to understand it.
Main thoughts:
1. If you had to write all these comments for us, they should likely also go into the code, at least portion of it that has any assumptions we did or reasons why we did something and similar.
2. Seems to me like I am just missing context to understand the crux of it. For example you say this command controls and emulates `tsc`. I can see it call `tsc`, but I don't get what you mean by controling and emulating it. I get that we have `tsconfig.node.json`, and I think I undrestand that `tsconfig.node.json` is here primarily for IDE, but here, in this call to `tsc`, we want to use settings from `tsconfig.node.json` but with some tweaks that we need to actually do the compliation here? And then you are tring to explain those tweaks and assumptions behind them? You continue the use of word `control` through the comment, and I feel like that should give me context I need to understand the rest, but it doesn't.
I also don't get the part with keeping contents of tsconfig.node.js in Haskell -> so we wouldn't have it on disk at all? How would you use them then, you would feed them to this command here via stdin? Can that even work and I would suspect then it makes more sense to jsut provide ocmpiler flags? I feel like I am again missing something here.

Reply:
@Martinsos I guess what Filip is trying to day by "emulate and control" - make sure that the compilation steps happens in the same say the TS LSP interprets the `tsconfig.node.json` in the IDE + modifying it to actually emit the file. 

**This command**  runs `tsc` with `tsconfig.node.json` + adding some options so the JS file is outputted.

**Motivation:** in order for us to know where the JS file will end up (so we can call it with `node`) we have to make sure that the user didn't modify the `tsconfig.node.json` in way that would mess up our pipeline. We do this by adding the `noEmit` and `outDir` options + we'll validate the `tsconfig.node.json` file.

Reply:
Ok tnx both -> @infomiho's explanation here helped + he explained it to me a bit more, so I think I get it now @sodic ! Here is what I understand, and how I would put it.

Here, in this function, we are using `tsc` to compile the main `.wasp.ts` file into JS file so we can execute it.
While `tsconfig.wasp.json` is the main config for compilation of `.wasp.ts` files (both for our purposes here and for their IDE (language server)), there are a couple of settings we want to ensure for here that are not in the tsconfig.json -> we want to specify that JS file has to be emitted, and we want to specify where it should be emitted, so that we can then later execute that JS file. So we add those as additional cmd args to `tsc`, overriding whatever is in `tsconfig.json`. Why didn't we write such tsconfig.json from the start? Well, because LS doesn't need it, and we can easily set these arguments here, so why enforce them earlier then we have to.
So TLDR: we set `noEmit` and `outDir` to make sure that JS file gets produced and in expected place, so we can further consume it.

That is it for that part, but there is another issue, which I think is a general issue, not specific to are we compiling with `tsc` or if LS is doing the compilation: How do we know what to put in `include` in tsconfig, when we don't konw the name of the `.wasp.ts` file? It is not fixed to `main.wasp.ts`, it can be anything instead, as long as it ends with `.wasp.ts`. So you are asking here if we should, at in this place in haskell code, where we already identified the `.wasp.ts` file, check if it indeed matches what is in `include` field in `tsconfig.wasp.json` and complain if that is not a match.
As I said, I would treat this as a separate issue, as it is a problem for both LS and for our `tsc` usage here.
One solution I see for it is indeed what you said -> check if these names match and complain otherwise.
Other solution I see is us hardcoding the file: we can say it has to be `main.wasp.ts` and that is it. Or `spec.wasp.ts`? Probalby `main.wasp.ts` is good enough name.

I actually like the idea of hardcoded name. We dont' need the flexibility + then we open the opportunity for multiple .wasp.ts files which we will want anyway, and you don't have to bother with validation or anything. We can just tell them it has to be `main.wasp.ts` and that is it.